### ISSUE TITLE
Add ability to create model elements in the model browser

### ISSUE THREAD

**Issue by marek-piirikivi** [gaphor/gaphor#2552]
### Is your feature request related to a problem?
When working with diagrams, sometimes it is necessary to reference some element in parameters that wont itself populate the diagram. One example would be activity diagram pin type selection or swimlane allocations. Additionally, more experienced modellers prefer to use model explorer extensively while modeling.

### Describe the solution you'd like

I would like to be able to right click in package/other model element and click "New SysML Element" and be able to see options such as "Block", "Activity", "Requirement", "ValueType" etc. Upon selecting the option, new model element would be created in the model explorer.

### Describe alternatives you've considered
Currently an element must be created in the diagram, then select an option to not delete elements that are deleted from the diagram and delete the element from the diagram. Then it can be used where necessary.

### Additional context
I am mostly interested in SysML profile, but if it makes sense, then similar behavior could also be extended to UML and other profiles.

**Comment by sz332** [comment]
I like the idea, especially that I realized that I am using direct modeling so many times a day. Let's start with some needs:

I tried to write some needs from the user (modeler) point of view. Let's discuss.

- As a modeler, I would like to create model element instances without creating a diagram so that I can concentrate on the model creation instead of the presentation part
- As a modeler, I would like to create instances from a well defined list so that I can choose the element I want to create
- As a modeler, I would like to be able to filter this well defined list (by name, by ...) so that I can select the element faster
- As a modeler, I would like to choose from both SysML and UML model elements 
- As a modeler, when I select a model element, I would like to enter the name of the instance
- As a modeler, when an instance of a model element is created, I would like define where it will be stored in the mode

**Comment by marek-piirikivi** [comment]
> * As a modeler, I would like to create model element instances without creating a diagram so that I can concentrate on the model creation instead of the presentation part

Yes. That's about sums up my needs nicely. You can control where and what model elements are going to live and do not have to leave it to a chance when they are first created on the diagram - only for you to then drag and move them to their correct locations. When combining it with #2545, you can pre-define some elements.

Another strong property I see with it is that it solidifies the idea of not recreating new elements in the diagram when you want to illustrate that the same element is used elsewhere as well - seems to be common mistake, at least what I have observed, that each diagram redefines elements.

> * As a modeler, I would like to create instances from a well defined list so that I can choose the element I want to create

Agreed.

> * As a modeler, I would like to be able to filter this well defined list (by name, by ...) so that I can select the element faster

I am not sure about that since well placed model element in the model structure should already be quite easily findable. "Easily findable" would be the desired property of the newly created elements.

I know that Gaphor already has a search option in the model browser, so I suppose this would be implicitly satisfied. I am not as seasoned as @sz332, so I can not comment on how useful this would be in everyday life, but I am not against filtering. @sz332, you may want to elaborate more on this - are you thinking that you should be able to filter based on the element types or relationships etc? I think it would deserve its own feature development and is not explicitly related to the ability to create elements in the model browser.

> * As a modeler, I would like to choose from both SysML and UML model elements

I have very little experience with UML myself, so I am not sure what is the workflow of those who use UML and whether it is an use-case. Most other tools I have seen indeed offer this feature to both SysML and UML elements. Again, it would not hurt to have it. Currently, many elements seen on SysML diagrams are technically UML elements as well.

> * As a modeler, when I select a model element, I would like to enter the name of the instance

Yes. Current default behavior on new diagram creation is to activate rename option. I think it should be applied to new elements as well. Additionally, I would also open property page for the newly created element once we have #2545.

> * As a modeler, when an instance of a model element is created, I would like define where it will be stored in the mode

I imagine that the creation would be similar to what it is with the diagrams; so you would start the process of the creation from the desired storage location. So, I would propose:

"As a modeler, I would like to start model element creation from its desired storage location in the model browser."

My additional view of the needs:

* As a modeler, I would like to be able to focus on the element relationships on the diagram and not to think about element definition.

Basically single-responsibility principle. You want to be able to think about one thing at the moment. A good analogy would be production line where you have your components already made and you just need to connect them - I see modeling similarly (not 100% applicable, but still illustrates the point).

**Comment by sz332** [comment]
About the filtering part: Imagine you have a hundred element types (class, block, stereotype, package, instancespecification, and so on). You want to find the element type you want to create the fastest way, and this is what I call filtering.

One way to implement this could be: press right mouse button on tree, the menu pops up. You go into New Element submenu, then drill down submenu by submenu until you find the element you are looking for.

Another way to implement this could be ... New Element submenu, then a dialog pops up, showing all the elements but having the ability to filter by entering the name in an input field on top of the menu.

Both solve the same problem.

**Comment by marek-piirikivi** [comment]
> Another way to implement this could be ... New Element submenu, then a dialog pops up, showing all the elements but having the ability to filter by entering the name in an input field on top of the menu.

I think this would work better than other alternative. The first alternative requires you to know the categories of the elements - which you may not always know.

> One way to implement this could be: press right mouse button on tree, the menu pops up. You go into New Element submenu, then drill down submenu by submenu until you find the element you are looking for.

The benefit of drilling would be that you do not have to touch your keyboard before the element is created. But since you probably want to rename it immediately after the creation - might as well touch the keyboard earlier. So, still I see that the second alternative is better.

If we also factor in the ability to open the menu using the keyboard only, then your flow can be quite fast. If I am not mistaken, it should be already possible to open right-click menu with keyboard (I need to confirm this).

> UML part: I would say that this feature should support every element type. So both UML elements and SysML elements. I see no reason to limit this to SysML.

Yes. I agree. However, maybe still filter based on the profile, because otherwise the profile loses its meaning - you would not want to see non-SysML options when SysML profile is selected, I suspect.

**Comment by sz332** [comment]
The only tricky situation for SysML would be really what to display. For example we use stereotypes a lot but profile and stereotype is not part of the SysML specification AFAIK, so would that mean that I need to switch in-between profiles just to be able to create stereotypes in my SysML model?

One solution to this problem could be a checkbox in the dialog (Show all elements [ ]), if this is checked, then all elements - independently from the selected profile - will be displayed in the list. Obviously we can make it more complex, but maybe it is enough. What do you think?

What do you think?

**Comment by marek-piirikivi** [comment]
I feel as if I had this discussion somewhere :thinking: about losing the entire concept of selecting profile - and just having options to turn on/off features. So, you may have your UML and SysML turned on at the same time and you would see both of their options. Or maybe SysML and RAAML only turned on etc. Or maybe you want to have all the features turned on at the same time so you would see all the possible options everywhere.

So, I support your checkbox idea and would extend it to have checkboxes for all the profiles and lose the dropdown - let the modellers configure what tools they want to access themselves.

**Comment by sz332** [comment]
In this case I would say simply display everything, and later extend this feature by the ability to display only those types which are included in the turned on "profiles".

**Comment by marek-piirikivi** [comment]
This "profile" kinda falls out of the scope already regards the ability to just create elements. I would propose that each modeling language (=profile) would define their own set of elements for now that can be created in the model explorer (along with any constraints - if applicable).

**Comment by sz332** [comment]
I like the idea of the declarative approach, I think that's the right way to do it.

**Comment by marek-piirikivi** [comment]
So, from what I have gathered from this discussion for the version 1 of the element creation from the model browser topic is as following:

- Upon right-clicking an element from the model browser, the model browser shall open a menu show category "New Element"
- "New Element" category shall show a list of new element creation options defined by a selected profile.
    -  The list shall show only standard conforming elements
         - Standard conforming elements are elements which can be contained by the selected model browser element
    - When SysML profile is selected, the list shall contain:
         - Activity
         - Actor
         - Block
         - Interaction
         - Requirement 
         - StateMachine
         - ValueType
     - When UML profile is selected, the list shall contain
         - Activity
         - Artifact
         - Class
         - Component 
         - DataType
         - Device
         - Interaction
         - Node
         - StateMachine
- Upon selecting new element, the model explorer shall add new element under the selected element with a default name "New << element type >>"
- Upon the new element creation, the model explorer shall select the element and initiate renaming activity.

@sz332 Please review. I am unsure of the RAAML and C4 options. We can also omit these for now until a new issue requesting these is opened. Since I have not used UML please also tell whether the list of UML options is useful.

**Comment by sz332** [comment]
I would add:

- Package: we have it currently as a separate menu item but we could move it here to be more consistent
- Interface (for UML) 
- InterfaceBlock (for SysML)
- Enumeration (both)
- Signal (both)
- Use Case (both)
- InstanceSpecification (both)
- Actor (both, not only UML)
- Profile (both)
- Stereotype (both)

**Comment by sz332** [comment]
One more extension: what if the model element on which right mouse button is pressed cannot own another elements (like, Requirement). What would be the behavior in this case?

**Comment by marek-piirikivi** [comment]
> One more extension: what if the model element on which right mouse button is pressed cannot own another elements (like, Requirement). What would be the behavior in this case?

You wouldn't see the option. That's the idea behind "Standard conforming elements are elements which can be contained by the selected model browser element". I should really write out the constraints in here as well.

**Comment by marek-piirikivi** [comment]
Small teaser for V1. All that is left to do is to dig bit deeper into what elements are well defined in the Gaphor and can even be shown as options; and additionally check containment constraints.

Uploading gaphor-2023-07-28_08.01.39.mp4…

**Comment by sz332** [comment]
![image](https://github.com/gaphor/gaphor/assets/8182138/29f01264-e4bf-403a-876e-cf529c16f4c4)

_A composite requirement can contain subrequirements in terms of a requirements hierarchy, specified using the UML
namespace containment mechanism. This relationship enables a complex requirement to be decomposed into its
containing child requirements. A composite requirement may state that the system shall do A and B and C, which can be
decomposed into the child requirements that the system shall do A, the system shall do B, and the system shall do C. An
entire specification can be decomposed into children requirements, which can be further decomposed into their children
to define the requirements hierarchy._

Now we might have a question about what can a requirement contain? Obviously, other requirements, but anything else? Well, to answer this question we can turn to the SySML standard.

![image](https://github.com/gaphor/gaphor/assets/8182138/6d011b86-3317-436e-9ded-791e62881977)

5_nestedclassifiers_are_requirements actually requires these constraints to be satisfied.

**Comment by mikekidner** [comment]
I agree that at times, quickly creating a bunch of elements via right-clicks would be great. It seems it would then be necessary to auto-generate the associated bdd's to align with Gaphors' diagram-first approach.

(Perhaps that the model browser structure could be improved, for instance I think that a composite relation on a bdd should cause the blocks to nest in the model browser - rather than having to relocate by hand)

In this suggested model-first behaviour, how would it handle keeping the diagrams and model in synch?

While the delete-element-not-on-diagram behaviour is shocking for some users, it does make sure that the diagrams define the model rather than being partial views of a database of elements. 

Deciding if Gaphor is diagram-first or model-first seems pretty fundamental in getting consistent use across the application.

**Comment by sz332** [comment]
I think this is a life-cycle management question. 

In reality, the rule does not say that a model element can exists only if there is a corresponding diagram element, it is not a structural constraint, but a behavior constraint: during the removal of a diagram element, if there are no other diagram elements point to the same model element, remove the model element as well. And also this behavior can be switched off even right now in the configuration.

On logical (conceptual) level, you simply cannot create a diagram element without creating a model element, so it cannot ever be a structural rule (unless they are being created in a transaction). 

Also think about importing: is there any value that during import we should decide about where something will be displayed? When importing thousands of model elements how much more things would become,  does it provide value, and how much more it complicates the process of the modeler? Does he have in-front create all the diagrams (or hope the modeling tool to figure out cleverly) and then go through each and every model element to be created and find it's place upfront or just press import button, enjoy the model elements being created, and gradually, step by step, create a diagram (serving as a view) on the model and add the model elements. 

So, adding model elements using a modeling tree with the current constraint in my opinion is still consistent, however I agree that we shall have a discussion (in a separate ticket) about what diagram-first approach really means, what are the implications, and given the current maturity of the tool, is this even relevant anymore?

**Comment by marek-piirikivi** [comment]
You can check out and try it through here https://github.com/gaphor/gaphor/pull/2585. I am open for any kind of feedback and also suggestions on what kind of elements should be first version show. I put the defaults how I saw what would be necessary. 

Also, I am looking for help about the containment. By default, all the new elements can only be owned by the package, so this is the only place that would allow to create the elements with the exception being requirements which also can be created under already existing requirements.

**Comment by shobeira** [comment]
Would be nice to be able to **search an element** in the model tree/model browser

**Comment by marek-piirikivi** [comment]
@shobeira You can already do that - the model browser has search bar when you start typing there. I suspect that you meant that in the new element creation? Is that so?

**Comment by shobeira** [comment]
> @shobeira You can already do that - the model browser has search bar when you start typing there. I suspect that you meant that in the new element creation? Is that so?

I am on Gaphor version 2.19.3 (flatpak) and not able to see the search bar. Would be great to have a search bar in the element creation process also.

**Comment by marek-piirikivi** [comment]
@shobeira The search bar appears when you click something in the model browser and just start typing. Otherwise it is hidden.

For the element creation I agree that we should also have it in there. I am not sure yet how much work would go into it to add it; but it might be a good follow-up issue.

**Comment by shobeira** [comment]
@marek-piirikivi In terms of user experience, it would be more practical to be visible by default and easily recognisable by the user. Otherwise, the user is not aware of the feature by heart.

**Comment by marek-piirikivi** [comment]
First version of this is now merged. @sz332 @shobeira I wouldn't mind to start collecting improvements if necessary. Maybe under a separate issue to be more explicit on what changes are planned. :slightly_smiling_face:

**Comment by sz332** [comment]
Created a new ticket: #2606 , this one can be closed in my opinon. @shobeira

**Comment by shobeira** [comment]
As the issue is opened by @marek-piirikivi, I am not able to close it.
